docs/RFCS/20200720_modern_passwords.md

```diff
@@ -0,0 +1,590 @@
+new storage method, we must prevent the creation of user accounts
+using the new method.
+
+Otherwise, it would be possible for version X nodes to mistakenly
+use the hashed details of the new method using the `bcrypt-crdb`
+algorithm, resulting in a potentially insecure situation.
+
+For this, we introduce a cluster version to gate usage of the new method.
+
+## Default storage method selection
+
+When a client provides a password in plaintext using `CREATE ROLE WITH
+PASSWORD`, the server must choose a storage method to compute
+`hashedPassword` in `system.users`.
+
+This is done by introducing a cluster setting
+`server.user_login.password_encryption`. This `password_encryption`
+setting has the same role as `password_encryption` in PostgreSQL: it
+determines the storage method to use for all new accounts, when
+the client configures an initial password in cleartext.
+
+In CockroachDB, the only two allowed values are `bcrypt-crdb` and
+`scram-sha-256`.
+
+We introduce a cluster setting as opposed to hard-coding the algorithm
+so as to facilitate tests and the generation of `system.users` tables
+compatible with previous versions of CockroachDB.
+
+## Storage method selection in WITH PASSWORD
+
+We want to make it possible to create password-based accounts without
+passing the cleartext password to the server.
+
+This is used e.g. in the CC management console to make the server
+blind to the cleartext password.
+
+For this we extend the SQL dialect as follows:
+
+- (baseline, already implemented:) `CREATE USER/ROLE WITH PASSWORD
+  'xxxx'`: the server understands `xxxx` as a cleartext password.
+
+  The server computes `system.users.hashedPassword` using the storage
+  method configured server-side via the setting
+  `server.user_login.password_encryption`.
+
+  The current value of `server.user_login.password_encryption` is stored
+  in the (new) `system.users.format` column.
+
+- `CREATE USER/ROLE WITH PASSWORD 'xxxx' ENCRYPTION 'yyyy'`: the
+  server understand `xxxx` as the direct `hashedPassword` value as
+  computed by the storage method `yyyy` client-side.
+
+  It stores `xxxx`
+  directly into `system.users.hashedPassword` and `yyyy` into the
+  (new) `system.users.format` column.
+
+Here is a v20.1-compatible example:
+
+- assume the current server configuration has `server.user_login.password_encryption = bcrypt-crdb`.
+- client sends `CREATE USER foo WITH PASSWORD 'abc'` (cleartext password).
+- server detects method `bcrypt-crdb` from cluster setting; selects salt and computes Bcrypt-10 hash
+- server stores `$2a$10$fGgWYzxv4UTXVTNzQTHEa.kX3pMNNE.mxxoSk1ZTF9MPZlLOkHxbK` into `system.users.hashedPassword`. 
+- server stored `bcrypt-crdb` into `system.users.format`.
+
+Here's an example using the new option, in a way compatible with v20.1 servers:
+
+- client sends `CREATE USER foo WITH PASSWORD '$2a$10$fGgWYzxv4UTXVTNzQTHEa.kX3pMNNE.mxxoSk1ZTF9MPZlLOkHxbK' ENCRYPTION 'bcrypt-crdb`
```

Comment:
The previous paragraph says that the server "selects salt and computes Bcrypt-10 hash". Can you describe here how the CC management console (or other users of this syntax) have access to this salt so they can correctly compute the hash? Will CC be able to use this method with the new scram stuff?